sno/working_copy/mysql.py

```diff
@@ -0,0 +1,327 @@
+                WHERE GC.table_schema=:table_schema AND GC.table_name=:table_name;
+            """
+            r = sess.execute(
+                spatial_ref_sys_sql,
+                {"table_schema": self.db_schema, "table_name": dataset.table_name},
+            )
+            mysql_spatial_ref_sys = list(r)
+
+            id_salt = f"{self.db_schema} {dataset.table_name} {self.get_db_tree()}"
+            schema = mysql_adapter.sqlserver_to_v2_schema(
+                mysql_table_info, mysql_spatial_ref_sys, id_salt
+            )
+            yield "schema.json", schema.to_column_dicts()
+
+    @classmethod
+    def try_align_schema_col(cls, old_col_dict, new_col_dict):
+        old_type = old_col_dict["dataType"]
+        new_type = new_col_dict["dataType"]
+
+        # Some types have to be approximated as other types in MySQL
+        if mysql_adapter.APPROXIMATED_TYPES.get(old_type) == new_type:
+            new_col_dict["dataType"] = new_type = old_type
+            for key in mysql_adapter.APPROXIMATED_TYPES_EXTRA_TYPE_INFO:
+                new_col_dict[key] = old_col_dict.get(key)
+
+        # Geometry types don't have to be approximated, except for the Z/M specifiers.
+        if old_type == "geometry" and new_type == "geometry":
+            old_gtype = old_col_dict.get("geometryType")
+            new_gtype = new_col_dict.get("geometryType")
+            if old_gtype and new_gtype and old_gtype != new_gtype:
+                if old_gtype.split(" ")[0] == new_gtype:
+                    new_col_dict["geometryType"] = new_gtype = old_gtype
+
+        return new_type == old_type
+
+    _UNSUPPORTED_META_ITEMS = (
+        "title",
+        "description",
+        "metadata/dataset.json",
+        "metadata.xml",
+    )
+
+    def _remove_hidden_meta_diffs(self, dataset, ds_meta_items, wc_meta_items):
+        super()._remove_hidden_meta_diffs(dataset, ds_meta_items, wc_meta_items)
+
+        # Nowhere to put these in SQL Server WC
+        for key in self._UNSUPPORTED_META_ITEMS:
+            if key in ds_meta_items:
+                del ds_meta_items[key]
+
+        # Diffing CRS is not yet supported.
+        for key in list(ds_meta_items.keys()):
+            if key.startswith("crs/"):
+                del ds_meta_items[key]
+
+    def _is_meta_update_supported(self, dataset_version, meta_diff):
+        """
+        Returns True if the given meta-diff is supported *without* dropping and rewriting the table.
+        (Any meta change is supported if we drop and rewrite the table, but of course it is less efficient).
+        meta_diff - DeltaDiff object containing the meta changes.
+        """
+        # For now, just always drop and rewrite.
+        return not meta_diff
+
+
+class GeometryType(UserDefinedType):
+    """UserDefinedType so that V2 geometry is adapted to MySQL binary format."""
+
+    # TODO: is "axis-order=long-lat" always the correct behaviour? It makes all the tests pass.
```

Comment:
Apparently it is, but I don't think that's relevant. MySQL wants to be told what to do on import from WKB / export to WKB - either lat-long, long-lat, or whatever the CRS standard officially defines (probably lat-long, at least for EPSG).

I think the order in which MySQL actually stores them is mostly hidden from us - MySQL knows which number is lat and which is long, and can import and export according to the rule we provide. That's what matters to us.

What I don't really know yet is why all our test data is apparently in long-lat ordering, which is apparently the opposite of the EPSG standard. I think there is a de facto standard that is the opposite of the actual standard, but I don't know if that is something we can rely on in general.

Reply:
I thought about this some more. Discussion of the basic issue here - you are doubtless more familiar with it than I am:

http://docs.geotools.org/latest/userguide/library/referencing/order.html

Internally, we store geometries using the GPKG format, which officially uses what it calls the "de facto" standard - an axis ordering of long-lat / East-North / XY. (It is XY in that it is across then up - but, all standards are XY in that we always call the first variable X). The GPKG format contains an embedded WKB. The WKB standard seems to leave open to interpretation what order the axes are in - either, they are in the "de facto" standard ordering (long-lat, East-North), or, they are in the order defined by the CRS (which would generally be the EPSG standard - lat-long, North-East). This second ordering seems to be mandated by ISO 19128:2005, which seems not to have been widely adopted since it is the opposite of the de facto standard and nobody seems to have thought of an actual way to gradually migrate every system so that it conforms to the new standard without breaking the de facto standard. 

Seems like all of the libraries we are using so far simply assume that "de facto" standard - although perhaps I should double check this - but MySQL is the first library which tries by default to conform to new standard, it assumes that the ordering from the CRS standard is used, which, since our WKB is from a GPKG, it definitely is not - instead it is definitely in GPKG / de-facto / long-lat / East-North ordering. That is why we need to explicitly tell MySQL which order it should actually expect in the WKB that we use to get data into and out of MySQL. 

So, I think I can delete this comment, or at least, add some background instead of a TODO. Maybe I should add something to the DATASETS_v2 doc to note that, since Kart stores everything using GPKG geometries, it always uses GPKG / "de-facto" / long-lat / East-North ordering. I think this is true of all the JSON / GEOJSON that Kart outputs too.